File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // Map from sorted-key → list of original strings sharing that key
        unordered_map<string, vector<string>> groups;
        for (const string& s : strs) {
            // Sorting the characters produces a canonical key for all anagrams
            string key = s;
            sort(key.begin(), key.end());
            groups[key].push_back(s);
        }
        // Flatten the map values into the result
        vector<vector<string>> result;
        result.reserve(groups.size());
        for (auto& [key, group] : groups) {
            result.push_back(move(group));
        }
        return result;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.go
func groupAnagrams(strs []string) [][]string {
    groups := make(map[string][]string)
    for _, s := range strs {
        // Sort characters to produce a canonical key for all anagrams
        runes := []rune(s)
        sort.Slice(runes, func(i, j int) bool {
            return runes[i] < runes[j]
        })
        key := string(runes)
        groups[key] = append(groups[key], s)
    }
    // Flatten map values into result slice
    result := make([][]string, 0, len(groups))
    for _, group := range groups {
        result = append(result, group)
    }
    return result
}